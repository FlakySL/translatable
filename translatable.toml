
locales_path = "./translatable/tests/environments/everything_valid/translations/"
