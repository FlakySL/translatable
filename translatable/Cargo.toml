[package]
name = "translatable"
description = "A robust internationalization solution for Rust featuring compile-time validation, ISO 639-1 compliance, and TOML-based translation management. "
repository = "https://github.com/FlakySL/translatable.rs"
license = "MIT OR Apache-2.0"
readme = "../README.md"
version = "0.1.0"
edition = "2024"
authors = ["Esteve Autet <esteve@memw.es>", "chikof"]
tags = ["i18n", "translations", "idioms", "languages", "internazionalization"]

[dependencies]
thiserror = "2.0.12"
translatable_proc = { path = "../translatable_proc" }

[dev-dependencies]
trybuild = "1.0.104"
