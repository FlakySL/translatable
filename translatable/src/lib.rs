use thiserror::Error;

// re export the macro in the main crate.
pub use translatable_proc::translation;

/// This error is used on results for the
/// translation procedural macro, the macro
/// will return a Result<Option<&'static str>, Error>,
/// when there is a dynamic expression to resolve.
///
/// For example, if the language is a dynamic expression
/// meaning it's not a literal &'static str, and it evaluates
/// on runtime, if the runtime evaluation is invalid because
/// the language does not match the ISO 639-1 specification
/// or something else, the translation macro will return an
/// Error::InvalidLanguage.
///
/// For more information on the possible errors read each
/// enum branch documentation.
#[derive(Error, Debug)]
pub enum Error {
    #[error("The language '{0}' is invalid.")]
    InvalidLanguage(String),

    #[error("The langauge '{0}' is not available for the path '{1}'")]
    LanguageNotAvailable(String, String)
}

/// This module is for internal usage, it's members
/// are not documented, and there is no support on
/// using it.
#[doc(hidden)]
pub mod internal {
    use std::collections::HashMap;

    #[doc(hidden)]
    pub enum NestingType {
        Object(HashMap<String, NestingType>),
        Translation(HashMap<String, String>)
    }

    #[doc(hidden)]
    impl NestingType {
        pub fn get_path(&self, path: Vec<&str>) -> Option<&HashMap<String, String>> {
            match self {
                Self::Object(nested) => {
                    let (first, rest) = path.split_first()?;

                    nested
                        .get(*first)
                        .and_then(|n| n.get_path(rest.to_vec()))
                },

                Self::Translation(translation) => {
                    if path.is_empty() {
                        return Some(translation)
                    }

                    None
                }
            }
        }
    }
}
