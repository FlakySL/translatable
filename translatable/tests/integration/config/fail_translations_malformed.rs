// the macro isn't filled because the expected
// failure is on configuration.

use translatable::{translation, Language};

fn fail_translations_malformed() {
    translation!(Language::ES, vec![]);
}
