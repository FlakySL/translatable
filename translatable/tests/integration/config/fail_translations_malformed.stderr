error: A nesting can only contain translation objects or other nestings
 --> tests/integration/config/fail_translations_malformed.rs
  |
  |     let _ = translation!(Language::ES, vec![""]);
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `translation` (in Nightly builds, run with -Z macro-backtrace for more info)
